### PR TITLE
Add tracking issue for Layout methods (and some API changes)

### PR DESCRIPTION
**Pull request by Amanieu**
These methods are already useful when used with the stable global allocator API (stabilized in #51241).

```rust
pub fn align_to(&self, align: usize) -> Result<Layout, LayoutErr>;
pub fn padding_needed_for(&self, align: usize) -> usize;
pub fn repeat(&self, n: usize) -> Result<(Layout, usize), LayoutErr>;
pub fn extend(&self, next: Layout) -> Result<(Layout, usize), LayoutErr>;
pub fn repeat_packed(&self, n: usize) -> Result<Layout, LayoutErr>;
pub fn extend_packed(&self, next: Layout) -> Result<Layout, LayoutErr>;
pub fn array<T>(n: usize) -> Result<Layout, LayoutErr>;
```

cc #32838 

r? @SimonSapin

### CODE REVIEW

File: src/libcore/alloc.rs [hunk 1]
```diff
@@ -169,7 +169,7 @@ impl Layout {
     /// Panics if the combination of `self.size()` and the given `align`
     /// violates the conditions listed in
     /// [`Layout::from_size_align`](#method.from_size_align).
-    #[unstable(feature = "allocator_api", issue = "32838")]
+    #[stable(feature = "alloc_layout_extra", since = "1.31.0")]
```
Comment (Centril) [inline src/libcore/alloc.rs]: Hmm... will this make it in time? Hasn't 1.31 gone into beta yet?

Comment (Amanieu) [reply src/libcore/alloc.rs]: I just used the version number of the current latest nightly, which is 1.31.

Comment (Centril) [reply src/libcore/alloc.rs]: cc @pietroalbini ^

**Comment by SimonSapin** [comment]
`repeat` and `extend`’s `Ok` return type is `(Layout, usize)`, where the `usize` is an offset.

`extend_packed` does the same, with docs saying:

> (The `offset` is always the same as `self.size()`; we use this signature out of convenience in matching the signature of `extend`.)

… but `repeat_packed` doesn’t, its `Ok` return type is just `Layout`.

This seems inconsistent for no reason, should we change `extend_packed` or `repeat_packed` to have the other’s return type?

**Comment by SimonSapin** [comment]
@rfcbot fcp merge

**Comment by SimonSapin** [comment]
@rfcbot concern return types consistency

For https://github.com/rust-lang/rust/pull/55366#issuecomment-433301851

File: src/libcore/alloc.rs (hunk 1 above)
Comment (pietroalbini) [reply src/libcore/alloc.rs]: 1.31 will go into beta in the weekend.

Comment (scottmcm) [reply src/libcore/alloc.rs]: FCP takes 10 days, so if merging needs to wait on FCP, this will be 1.32.

**Comment by SimonSapin** [comment]
@rfcbot concern compat with repr(C)

For example:

```rust
#[repr(C)]
struct Pair<A, B> {
   a: A,
   b: B,
}

assert_eq!(Layout::new::<Pair<Foo, Bar>>(),
           Layout::new::<Foo>().extend(Layout::new::<Bar>()))
```

Does this hold for any types `Foo` and `Bar`? Do we want to guarantee it’ll always hold? If so, let’s document that guarantee.

Similarly for `repr(C, packed)` and `extend_packed`.

Note that `Layout::array` is already documented as compatible with `[T; n]`.

**Comment by Centril** [comment]
> Does this hold for any types Foo and Bar? Do we want to guarantee it’ll always hold? If so, let’s document that guarantee.
>
> Similarly for repr(C, packed) and extend_packed.

cc @rust-lang/lang

**Comment by SimonSapin** [comment]
I suspected that but wasn’t sure how C struct layout is defined. Could you add sentence or two about this in the doc-comments of the relevant methods?

**Comment by Amanieu** [comment]
> This seems inconsistent for no reason, should we change extend_packed or repeat_packed to have the other’s return type?

I feel that we definitely need to change *something*. My preference would be to change`extend_packed` to just return a plain `Layout` for its `Ok`. I don't think that we gain anything by returning tuple, since I don't see people switching between the normal and packed layouts on a whim.

**Comment by SimonSapin** [comment]
@Amanieu I personally agree with changing `extern_packed` to return `Layout` without a tuple.

@rust-lang/libs Any thoughts on the **API change** proposed above?

@rfcbot resolve compat with repr(C)

**Comment by Amanieu** [comment]
The main use case is to construct complex allocation layouts for use with the stable global allocator API. For example:
- [`std::collections::HashMap`](https://github.com/rust-lang/rust/blob/master/src/libstd/collections/hash/table.rs#L657)
- [`hashbrown`](https://github.com/Amanieu/hashbrown/blob/master/src/raw/mod.rs#L134)
- Not exactly an example of use, but it would be very useful in [`crossbeam-skiplist`](https://github.com/crossbeam-rs/crossbeam-skiplist/blob/master/src/base.rs#L99)

**Comment by Amanieu** [comment]
Regarding `align_to`, I agree with you, this should probably be changed to return a `Result` like the other `Layout` methods.

**Comment by SimonSapin** [comment]
> these afaik haven't gone through a great deal of review

It sounds like we’re on a good start to do the review in this thread :) And if we’re still unsure about some some methods we can leave them for now and stabilize the rest.

> The `usize` return value with offsets seems a little odd to me, is that a value that's really going to be stored at runtime somewhere and then later looked up to actually access values?

I’d expect the offsets not to be stored. (Growing `HashMap` by a couple `usize`s for example would be non-trivial cost.) Instead they can be re-computed every time they’re needed. Since `size_of::<T>()` is constant after monomorphization, much of a that computation can be constant-folded by the optimizer.

**Comment by alexcrichton** [comment]
I suppose I'm not really comfortable myself proposing stablization without motivation mentioned in the description *and also* doing an API review at the same time.

The `align_to` and `array` methods I think are probably fine to stabilize (with returning a result from `align_to`). The others have the uncomfortable `usize` return value and also seem largely less motivated other than `extend` perhaps.

Is there a way perhaps the `usize` return value could be tweaked to be more obvious or communicated via a different means?

**Comment by Amanieu** [comment]
@alexcrichton I personally feel that the `extend` method is also important to stabilize, with the `usize` parameter. In both places where it is currently used (`hashbrown` and `std::collections::HashMap`) this is used to effectively allocate 2 separate arrays with a single call to the allocator. The offset is used to get the start address of the 2nd array after the allocation.

For `std::collections::HashMap`, the offset to the data array is re-calculated every time a method is called. For `hashbrown`, a separate data pointer is used since it can be left dangling when the table is empty.

**Comment by alexcrichton** [comment]
Ok well at the very least, I would want to restrict a consideration of stabilization to only `align_to`, `array`, and `extend`. The `padding_needed_for`, `repeat_packed`, and `extend_packed` stabilizations seem over ambitious and largely "they were already there and why not stabilize".

I'm still not sure what to do about `extend`. I continue to be somewhat uncomfortable that this is only coming up now *as stabilization is proposed to happen* and the rationale is "I've used it here, it was convenient, let's stabilize it as-is".

**Comment by SimonSapin** [comment]
I think that `extend` is fundamental to the principle of having somwhat-high-level APIs for computing non-trivial layouts, rather than making users compute `usize` values then calling `from_size_align`.

`padding_needed_for` is not particularly useful *if* `extend` exists.

I don’t know when `*_packed` would be used, but then the same applies to `#[repr(packed)]`.

**Comment by alexcrichton** [comment]
FWIW these sorts of comments are what I would expect to be, for example, in the description of this PR, the description of an FCP to stabilize, etc. Right now these comments are buried in a discussion that is quite long at this point and hard to extract, but the comments provide good rationale for stabilizing these methods. 

Procedurally I would prefer at this point if we open a dedicated tracking issue for the proposal to stabilize here. Such a tracking issue would contain rationale, use cases, etc, along with the proposed-to-be-stabilized interface. Other issues mentioning these APIs could then also get a comment indicating that this tracking issue is proposing stabilization of a few APIs.

I'm gonna go ahead and register a formal concern for this PR:

@rfcbot concern not-ready-yet

**Comment by Amanieu** [comment]
I have changed the PR to just split the `Layout` methods into a separate tracking issue (#55724).

**Comment by alexcrichton** [comment]
Thanks for the update! r=me but it looks like there's some minor build failures

@bors: delegate+

**Comment by Amanieu** [comment]
The failure seems to be unrelated to this PR?

@bors retry

**Comment by kennytm** [comment]
The nomicon failure:

```
[01:03:36] ---- /checkout/src/doc/nomicon/src/vec-final.md - The_Final_Code (line 3) stdout ----
[01:03:36] error[E0658]: use of unstable library feature 'alloc_layout_extra' (see issue #55724)
[01:03:36]   --> /checkout/src/doc/nomicon/src/vec-final.md:36:40
[01:03:36]    |
[01:03:36] 34 |                 let ptr = Global.alloc(Layout::array::<T>(1).unwrap());
[01:03:36]    |                                        ^^^^^^^^^^^^^^^^^^
[01:03:36]    |
[01:03:36]    = help: add #![feature(alloc_layout_extra)] to the crate attributes to enable
[01:03:36] 
[01:03:36] error[E0658]: use of unstable library feature 'alloc_layout_extra' (see issue #55724)
[01:03:36]   --> /checkout/src/doc/nomicon/src/vec-final.md:42:42
[01:03:36]    |
[01:03:36] 40 |                                          Layout::array::<T>(self.cap).unwrap(),
[01:03:36]    |                                          ^^^^^^^^^^^^^^^^^^
[01:03:36]    |
[01:03:36]    = help: add #![feature(alloc_layout_extra)] to the crate attributes to enable
[01:03:36] 
[01:03:36] error[E0658]: use of unstable library feature 'alloc_layout_extra' (see issue #55724)
[01:03:36]   --> /checkout/src/doc/nomicon/src/vec-final.md:43:42
[01:03:36]    |
[01:03:36] 41 |                                          Layout::array::<T>(new_cap).unwrap().size());
[01:03:36]    |                                          ^^^^^^^^^^^^^^^^^^
[01:03:36]    |
[01:03:36]    = help: add #![feature(alloc_layout_extra)] to the crate attributes to enable
[01:03:36] 
[01:03:36] error[E0658]: use of unstable library feature 'alloc_layout_extra' (see issue #55724)
[01:03:36]   --> /checkout/src/doc/nomicon/src/vec-final.md:69:32
[01:03:36]    |
[01:03:36] 67 |                                Layout::array::<T>(self.cap).unwrap());
[01:03:36]    |                                ^^^^^^^^^^^^^^^^^^
[01:03:36]    |
[01:03:36]    = help: add #![feature(alloc_layout_extra)] to the crate attributes to enable
```

**Pull request merged**